### PR TITLE
GEOMESA-2906 Create GeoAvroRecordSetWriter

### PR DESCRIPTION
**Pull request by jnh5y**
* Implements GeoAvroRecordSetWriter.
* Uses properties or existing SimpleFeature(Type) information.

Signed-off-by: Jim Hughes <jnh5y@ccri.com>

### CODE REVIEW

**Comment by jnh5y** [comment]
@elahrvivaz I tossed this up as a 'draft' PR so that we can identify what are next follow-on steps.

In terms of things to resolve before we merge this PR:

1.  Deal with raw(i) vs raw(i+1) offset!
(I'm surprised that a unit test didn't break when I changed this!)
2.  Discuss if the useProvidedFid should be true!
(IMO, seems like we ought to use whatever FID is provided; it is worth discussing.)

**Comment by jnh5y** [comment]
In terms of future work, I'd like to suggest:
* A property for field names to exclude
* Add tests for Wkb (since I skated over them here)
* Add property to select column as FID
* Handling for List and Map types:)
* Implement GeoAvro Reader
** Would write WKB as WKT (maybe optionally?)

Are there other ideas? 

Do any of these ideas need to be implemented in the first pass?

File: geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/SimpleFeatureRecordConverter.scala [hunk 1]
```diff
@@ -34,17 +36,93 @@ object SimpleFeatureRecordConverter {
...
+        }
+        // The default geometry is set as the first column name.
+        defaultGeometryColumn.foreach(builder.setDefaultGeometry)
+        typeName.foreach(builder.setName)
+
+        builder.buildFeatureType()
+    }
+  }
+
+  private val geometryTypeMap = scala.collection.Map(
```
Comment (elahrvivaz) [inline geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/SimpleFeatureRecordConverter.scala:93]: is this approach going to be case sensitive? do we care? (not sure how we handle that in normal sft parsing)

File: geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/SimpleFeatureRecordConverter.scala [hunk 2]
```diff
@@ -87,21 +157,32 @@ object SimpleFeatureRecordConverter {
...
     }
 
     override def convert(feature: Record): SimpleFeature = {
       val raw = feature.getValues
       val values = Array.ofDim[AnyRef](converters.length)
       var i = 0
       while (i < converters.length) {
-        values(i) = converters(i).toAttribute(raw(i + 1))
+        // Why is this raw(i+1)?
+        values(i) = converters(i).toAttribute(raw(i))
```
Comment (elahrvivaz) [inline geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/SimpleFeatureRecordConverter.scala:169]: i think this still needs to be +1 to skip over the fid, no?

File: geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/package.scala [hunk 3]
```diff
@@ -13,5 +13,11 @@ package object records {
   object GeometryEncoding extends Enumeration {
     type GeometryEncoding = Value
     val Wkt, Wkb = Value
+    def apply(encodingString: String): GeometryEncoding = {
+      encodingString match {
+        case "Wkb" => GeometryEncoding.Wkb
+        case "Wkt" => GeometryEncoding.Wkt
```
Comment (elahrvivaz) [inline geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/package.scala:19]: might be good to lowercase match, and add a `case _ => throw ` for clarity

File: pom.xml [hunk 4]
```diff
@@ -144,11 +144,28 @@
...
             <dependency>
                 <groupId>org.apache.nifi</groupId>
                 <artifactId>nifi-schema-registry-service-api</artifactId>
                 <version>${nifi.version}</version>
             </dependency>
+            <dependency>
+                <groupId>org.apache.nifi</groupId>
+                <artifactId>nifi-record-serialization-services</artifactId>
+                <version>${nifi.version}</version>
+                <scope>test</scope>
```
Comment (elahrvivaz) [inline pom.xml:161]: we should double check how the scope of these impacts things - nifi processors are not entirely intuitive with the expected scope of dependencies in a nar.

File: geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/SimpleFeatureRecordConverter.scala (hunk 1 above)
Comment (jnh5y) [reply geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/SimpleFeatureRecordConverter.scala:93]: Yes.  No, I don't think so.  Here's how we handle it: https://github.com/locationtech/geomesa/blob/main/geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/sft/SimpleFeatureSpec.scala#L274-L283
;)

File: geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/SimpleFeatureRecordConverter.scala (hunk 2 above)
Comment (jnh5y) [reply geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/SimpleFeatureRecordConverter.scala:169]: In the unit test I wrote, it was breaking with it as +1....

File: pom.xml (hunk 4 above)
Comment (jnh5y) [reply pom.xml:161]: Ok, thoughts about how to do that well?

File: geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/GeoAvroRecordSetWriterFactory.scala [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+@Tags(Array("avro", "geoavro", "result", "set", "recordset", "record", "writer", "serializer", "row"))
+@CapabilityDescription("Writes the contents of a RecordSet as GeoAvro which AvroToPutGeoMesa* Processors can use.")
+class GeoAvroRecordSetWriterFactory extends AbstractControllerService with RecordSetWriterFactory {
+  // NB: This is the same as what the InheritSchemaFromRecord Strategy does
+  override def getSchema(map: util.Map[String, String], recordSchema: RecordSchema): RecordSchema = recordSchema
+
+  override def createWriter(componentLog: ComponentLog,
+                            recordSchema: RecordSchema,
+                            outputStream: OutputStream,
+                            map: util.Map[String, String]): GeoAvroRecordSetWriter = {
```
[the remark this answers is not in the document]
Comment (jnh5y) [reply geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/GeoAvroRecordSetWriterFactory.scala:31]: Looks like it is the FlowFile attributes?  I'd imagine that any NiFi Expressions have already been evaluated.

File: geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/GeoAvroRecordSetWriterFactory.scala [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+object GeoAvroRecordSetWriterFactory {
+  val GEOMETRY_COLUMNS: PropertyDescriptor = new PropertyDescriptor.Builder()
+    .name("Geometry Columns")
+    .description("Comma-separated list of columns with geometries with type " +
+      "(and optionally the format, (Wkt, Wkb)).  " +
+      "Example: position:Point,line:LineString.  " +
+      "Note the first field is used as a the default geometry.")
+    .expressionLanguageSupported(ExpressionLanguageScope.NONE)
```
[the remark this answers is not in the document]
Comment (jnh5y) [reply geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/GeoAvroRecordSetWriterFactory.scala]: Changed this one to VariableRegistry.

File: geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/package.scala (hunk 3 above)
Comment (jnh5y) [reply geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/package.scala]: Good idea.  Can you see if you like how I've updated it?

File: geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/SimpleFeatureRecordConverter.scala (hunk 1 above)
Comment (elahrvivaz) [reply geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/SimpleFeatureRecordConverter.scala:93]: i guess we're more lenient with all the other types but not geoms... eh

File: geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/SimpleFeatureRecordConverter.scala (hunk 2 above)
Comment (elahrvivaz) [reply geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/SimpleFeatureRecordConverter.scala:169]: maybe the by name case is different than the by attribute case? seems odd that the unit test hasn't changed though... can you dig into the old unit test and figure out what's going on?

File: geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/package.scala (hunk 3 above)
Comment (elahrvivaz) [reply geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/package.scala]: looks good, thanks.

File: pom.xml (hunk 4 above)
Comment (elahrvivaz) [reply pom.xml:161]: i was thinking just compare the dependency tree from before and after and make sure none of the existing dependencies changed scope

File: geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/GeoAvroRecordSetWriterFactory.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (elahrvivaz) [reply geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/GeoAvroRecordSetWriterFactory.scala:31]: maybe we could rename the param to make it clearer, although if the super method has decent javadoc maybe it doesn't matter.

**Review by michaelchehreh-ccri — COMMENTED** [review]
I had to change these around to get it to work. Tests still passed with the change, which is odd. I was hoping to debug the test but ran out of time last week. Not certain I'll have enough free time this week to do it myself

```scala
            val writer = buildWriter(reader.getSft, sft, matchMode)
```

**Comment by michaelchehreh-ccri** [comment]
> I had to change [this line](https://github.com/geomesa/geomesa-nifi/blob/d749792869111e2b08b899499ec0fba66cdc706e/geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/AvroIngestProcessor.scala#L137) around to get it to work. Tests still passed with the change, which is odd. I was hoping to debug the test but ran out of time last week. Not certain I'll have enough free time this week to do it myself
> 
> ```scala
>             val writer = buildWriter(sft, reader.getSft, matchMode)
> ```
> 
> became
> 
> ```scala
>             val writer = buildWriter(reader.getSft, sft, matchMode)
> ```

@jnh5y Seems like the test succeeds because the order of the attributes that remain are the same, so the mappings (`0 -> 0` `1 -> 1` `4 -> 4`) are commutative. Below are the sfts
```
reader.getSft
  .getAttributeDescriptors.asScala.map(ad => (ad.getLocalName, ad.getType.getBinding.getCanonicalName))
 0 = {Tuple2@10028} "(fid,java.lang.Integer)"
 1 = {Tuple2@10029} "(name,java.lang.String)"
 2 = {Tuple2@10030} "(friends,java.lang.String)"
 3 = {Tuple2@10031} "(lastseen,java.util.Date)"
 4 = {Tuple2@10032} "(geom,org.locationtech.jts.geom.Point)"

sft
  .getAttributeDescriptors.asScala.map(ad => (ad.getLocalName, ad.getType.getBinding.getCanonicalName))
 0 = {Tuple2@10039} "(fid,java.lang.Integer)"
 1 = {Tuple2@10040} "(name,java.lang.String)"
 2 = {Tuple2@10041} "(age,java.lang.Integer)"
 3 = {Tuple2@10042} "(dtg,java.util.Date)"
 4 = {Tuple2@10043} "(geom,org.locationtech.jts.geom.Point)"
```
\
So after changing the order of the attributes in the application.conf it fails w/o changing the order in `buildWriter(...)`
```
geomesa = {
  sfts = {
    "example" = {
      attributes = [
        { name = "fid",  type = "Integer",         index = false }
        { name = "age",  type = "Integer" }
        { name = "dtg",  type = "Date",   default = true }
        { name = "geom", type = "Point",  srid = 4326, default = true }
        { name = "name", type = "String"  }
      ]
    }
  }
...
```

File: geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/AvroIngestProcessor.scala [hunk 7]
```diff
@@ -47,13 +47,13 @@ object AvroIngestProcessor {
         // Schemas match, so use a regular writer.
         // TODO:  Discuss if the useProvidedFid should be true!
         (fw: FeatureWriter[SimpleFeatureType, SimpleFeature], sf: SimpleFeature) =>
-          FeatureUtils.write(fw, sf, true)
+          FeatureUtils.write(fw, sf)
       case Failure(error) =>
         if (matchMode == LenientMatch) {
           val sfConverter = convert(sft1, sft2)
```
Comment (michaelchehreh-ccri) [inline geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/AvroIngestProcessor.scala:53]: It's possible you could switch the order somewhere in here instead...? Didn't do the most thorough check (logically I'm not convinced it should work) but `testAvroIngestByName` ran fine with changing this to `convert(sft2, sft1)`

Also, I forgot to run **all** tests with the changes I suggested in the other thread, and turns out the changes to the sft attribute order messes up another test. Seems like I might have to let you tackle the fixing part of this :smile:

**Comment by elahrvivaz** [comment]
replaced with #86

**Pull request closed without merging**